lib/extract.ts

```diff
@@ -0,0 +1,107 @@
+import { promises } from 'fs';
+import { join, extname, basename, relative } from 'path';
+import { FilePath } from './types';
+import { DECOMPRESSING_CONCURRENCY_LEVEL, EXTRACTED_DIR_FLAG } from './common';
+import { debug } from './debug';
+
+const pMap = require('p-map');
+const AdmZip = require('adm-zip');
+const tar = require('tar');
+
+const { mkdir } = promises;
+
+const zipFormats = ['.zip', '.zipx'];
+const tarFormats = ['.tar', '.gz', '.tgz'];
+
+interface ExtractionHandler {
+  (path: FilePath): void;
+}
+
+async function handleExtraction(
+  path: FilePath,
+  temporaryDir: FilePath,
+  childArchiveHandler: ExtractionHandler,
+) {
+  const extractionTarget: FilePath = join(
+    temporaryDir,
+    path.includes(temporaryDir)
+      ? relative(temporaryDir, `${path}${EXTRACTED_DIR_FLAG}`)
+      : basename(path),
+  );
+
+  await mkdir(extractionTarget, { recursive: true });
+
+  if (isTar(path)) {
+    await tar.x({
+      file: path,
+      cwd: extractionTarget,
+      sync: true,
+      onentry: (entry: any) => {
+        const childAbsolutePath = join(extractionTarget, entry.path);
+
+        if (isArchive(childAbsolutePath)) {
+          childArchiveHandler(childAbsolutePath);
+        }
+      },
+    });
+  }
+
```

Comment:
I guess we want the first if to be an early exit, or use `else if` to indicate that it's one or the other, not both.

Reply:
(In this case I would say `else if` or `switch` since we are branching out in the execution tree)

Reply:
This should be some form of pattern matching, unfortunately we can not use a `switch`. I updated to a specific else branch for the zips. Thanks for spotting this!